pkg/controller/pod.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	kubeClientSet "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+//Struct for the reconciler
+type ReconcilePod struct {
+	client                client.Client
+	kubeClient            kubeClientSet.Interface
+	azVolumeClient        azClientSet.Interface
+	namespace             string
+	controllerSharedState *SharedState
+}
+
+// Implement reconcile.Reconciler so the controller can reconcile objects
+var _ reconcile.Reconciler = &ReconcilePod{}
+
+func (r *ReconcilePod) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var pod corev1.Pod
+
+	podKey := getQualifiedName(request.Namespace, request.Name)
+	if err := r.client.Get(ctx, request.NamespacedName, &pod); err != nil {
+		// if the pod has been deleted, remove entry from podToClaimsMap and claimToPodMap
+		if errors.IsNotFound(err) {
+			r.controllerSharedState.deletePod(podKey)
+			return reconcile.Result{}, nil
+		}
+		klog.Errorf("failed to get Pod (%s): %v", podKey, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// update the shared map
+	r.controllerSharedState.addPod(&pod)
+
+	// create replica batches
```

Comment:
Do we have sufficient information in the Pod at creation time to being creating replicas or should we be waiting for the Pod to be running (or containers to be ready)?

Reply:
We only watch for update events where pod's status  phase transitions to `Running`.

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/2260a7332293f384c962089db81bbdb0940c79cf/pkg/controller/pod.go#L177-L184